src/Psalm/Checker/Statements/Expression/Call/MethodCallChecker.php

```diff
@@ -542,5 +544,57 @@ public static function analyze(
 
             $context->vars_in_scope[$var_id] = $class_type;
         }
+
+        // If a magic getter or setter is used to access a property on a class that is not defined
+        // but a `@property` annotation for the property exists, throw an error. If no `@property`
+        // annotation exists, it's not an error because you're allowed to make magic getters and
+        // setters do crazy things.
+        if ($stmt->var->inferredType && is_string($stmt->name)) {
```

Comment:
This can start after `checkMethodNotDeprecated`. You probably want to create a private method that has this logic too.